File: backend/app/schemas/quotation.py
from pydantic import BaseModel, Field, validator
from typing import Optional, List
from datetime import datetime
from decimal import Decimal

class QuotationAttachmentBase(BaseModel):
    file_name: str
    file_type: str
    description: Optional[str] = None

class QuotationAttachmentCreate(QuotationAttachmentBase):
    file_url: str

class QuotationAttachment(QuotationAttachmentBase):
    id: int
    quotation_id: int
    file_url: str
    created_at: datetime

    class Config:
        from_attributes = True

class QuotationBase(BaseModel):
    unit_price: Decimal = Field(..., ge=0, decimal_places=2)
    total_price: Decimal = Field(..., ge=0, decimal_places=2)
    currency: str = Field(default="INR")
    delivery_time: int = Field(..., gt=0)
    delivery_terms: str
    warranty_period: str
    payment_terms: str
    validity_period: int = Field(..., gt=0)
    technical_specifications: Optional[str] = None
    terms_and_conditions: Optional[str] = None
    notes: Optional[str] = None

    @validator('currency')
    def validate_currency(cls, v):
        allowed_currencies = ['INR', 'USD', 'EUR', 'GBP']
        if v not in allowed_currencies:
            raise ValueError(f'Currency must be one of {allowed_currencies}')
        return v

class QuotationCreate(QuotationBase):
    rfq_id: int
    attachments: Optional[List[QuotationAttachmentCreate]] = []

class QuotationUpdate(QuotationBase):
    unit_price: Optional[Decimal] = None
    total_price: Optional[Decimal] = None
    currency: Optional[str] = None
    delivery_time: Optional[int] = None
    delivery_terms: Optional[str] = None
    warranty_period: Optional[str] = None
    payment_terms: Optional[str] = None
    validity_period: Optional[int] = None

class Quotation(QuotationBase):
    id: int
    rfq_id: int
    supplier_id: int
    status: str
    created_at: datetime
    updated_at: datetime
    submitted_at: Optional[datetime] = None
    attachments: List[QuotationAttachment] = []

    class Config:
        from_attributes = True

File: frontend/src/lib/api/quotation.ts
import { supabase } from '@/lib/supabase'

export interface Quotation {
  id: number
  rfq_id: number
  supplier_id: number
  unit_price: number
  total_price: number
  currency: string
  delivery_time: number
  delivery_terms: string
  warranty_period: string
  payment_terms: string
  validity_period: number
  technical_specifications?: string
  terms_and_conditions?: string
  notes?: string
  status: string
  created_at: string
  updated_at: string
  submitted_at?: string
  attachments: QuotationAttachment[]
}

export interface QuotationAttachment {
  id: number
  quotation_id: number
  file_name: string
  file_url: string
  file_type: string
  description?: string
  created_at: string
}

export async function createQuotation(formData: FormData) {
  const { data: session } = await supabase.auth.getSession()
  if (!session) {
    throw new Error('Not authenticated')
  }

  const response = await fetch('/api/v1/quotations', {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${session.access_token}`
    },
    body: formData
  })

  if (!response.ok) {
    const error = await response.json()
    throw new Error(error.detail || 'Failed to create quotation')
  }

  return response.json()
}

export async function updateQuotation(quotationId: number, data: Partial<Quotation>) {
  const { data: session } = await supabase.auth.getSession()
  if (!session) {
    throw new Error('Not authenticated')
  }

  const response = await fetch(`/api/v1/quotations/${quotationId}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${session.access_token}`
    },
    body: JSON.stringify(data)
  })

  if (!response.ok) {
    const error = await response.json()
    throw new Error(error.detail || 'Failed to update quotation')
  }

  return response.json()
}

export async function submitQuotation(quotationId: number) {
  const { data: session } = await supabase.auth.getSession()
  if (!session) {
    throw new Error('Not authenticated')
  }

  const response = await fetch(`/api/v1/quotations/${quotationId}/submit`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${session.access_token}`
    }
  })

  if (!response.ok) {
    const error = await response.json()
    throw new Error(error.detail || 'Failed to submit quotation')
  }

  return response.json()
}

export async function acceptQuotation(quotationId: number) {
  const { data: session } = await supabase.auth.getSession()
  if (!session) {
    throw new Error('Not authenticated')
  }

  const response = await fetch(`/api/v1/quotations/${quotationId}/accept`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${session.access_token}`
    }
  })

  if (!response.ok) {
    const error = await response.json()
    throw new Error(error.detail || 'Failed to accept quotation')
  }

  return response.json()
}

export async function rejectQuotation(quotationId: number) {
  const { data: session } = await supabase.auth.getSession()
  if (!session) {
    throw new Error('Not authenticated')
  }

  const response = await fetch(`/api/v1/quotations/${quotationId}/reject`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${session.access_token}`
    }
  })

  if (!response.ok) {
    const error = await response.json()
    throw new Error(error.detail || 'Failed to reject quotation')
  }

  return response.json()
}

export async function getRFQQuotations(rfqId: number) {
  const { data: session } = await supabase.auth.getSession()
  if (!session) {
    throw new Error('Not authenticated')
  }

  const response = await fetch(`/api/v1/quotations/rfq/${rfqId}`, {
    headers: {
      'Authorization': `Bearer ${session.access_token}`
    }
  })

  if (!response.ok) {
    const error = await response.json()
    throw new Error(error.detail || 'Failed to fetch quotations')
  }

  return response.json() as Promise<Quotation[]>
}

export async function getSupplierQuotations() {
  const { data: session } = await supabase.auth.getSession()
  if (!session) {
    throw new Error('Not authenticated')
  }

  const response = await fetch('/api/v1/quotations/supplier', {
    headers: {
      'Authorization': `Bearer ${session.access_token}`
    }
  })

  if (!response.ok) {
    const error = await response.json()
    throw new Error(error.detail || 'Failed to fetch quotations')
  }

  return response.json() as Promise<Quotation[]>
}
